File: perf/benchmarks/apps/mandelbrot/mandelbrot.go
// Mandelbrot — application-style numerical benchmark (Go).
package main

import (
	"fmt"
	"os"
	"strconv"
	"time"
)

func mandelbrot(width, maxIter int) int64 {
	invW := 2.0 / float64(width)
	var total int64
	for py := 0; py < width; py++ {
		ci := float64(py)*invW - 1.0
		for px := 0; px < width; px++ {
			cr := float64(px)*invW - 1.5
			zr, zi := 0.0, 0.0
			n := 0
			for n < maxIter {
				zr2 := zr * zr
				zi2 := zi * zi
				if zr2+zi2 > 4.0 {
					break
				}
				zi = 2.0*zr*zi + ci
				zr = zr2 - zi2 + cr
				n++
			}
			total += int64(n)
		}
	}
	return total
}

func main() {
	width := 600
	maxIter := 200
	if len(os.Args) > 1 {
		v, _ := strconv.Atoi(os.Args[1])
		width = v
	}
	if len(os.Args) > 2 {
		v, _ := strconv.Atoi(os.Args[2])
		maxIter = v
	}
	t0 := time.Now()
	cs := mandelbrot(width, maxIter)
	ms := float64(time.Since(t0).Nanoseconds()) / 1_000_000.0
	fmt.Printf("CHECKSUM:%d\n", cs)
	fmt.Printf("ELAPSED_MS:%.6f\n", ms)
}

File: perf/benchmarks/apps/mandelbrot/mandelbrot.rs
// Mandelbrot — application-style numerical benchmark (Rust).
use std::env;
use std::time::Instant;

fn mandelbrot(width: i32, max_iter: i32) -> i64 {
    let inv_w = 2.0 / width as f64;
    let mut total: i64 = 0;
    for py in 0..width {
        let ci = py as f64 * inv_w - 1.0;
        for px in 0..width {
            let cr = px as f64 * inv_w - 1.5;
            let mut zr: f64 = 0.0;
            let mut zi: f64 = 0.0;
            let mut n: i32 = 0;
            while n < max_iter {
                let zr2 = zr * zr;
                let zi2 = zi * zi;
                if zr2 + zi2 > 4.0 { break; }
                zi = 2.0 * zr * zi + ci;
                zr = zr2 - zi2 + cr;
                n += 1;
            }
            total += n as i64;
        }
    }
    total
}

fn main() {
    let args: Vec<String> = env::args().collect();
    let width: i32 = if args.len() > 1 { args[1].parse().unwrap_or(600) } else { 600 };
    let max_iter: i32 = if args.len() > 2 { args[2].parse().unwrap_or(200) } else { 200 };
    let t0 = Instant::now();
    let cs = mandelbrot(width, max_iter);
    let ms = t0.elapsed().as_nanos() as f64 / 1_000_000.0;
    println!("CHECKSUM:{}", cs);
    println!("ELAPSED_MS:{:.6}", ms);
}
